src/caffe/layers/lifted_struct_similarity_softmax_layer.cpp

```diff
@@ -0,0 +1,219 @@
+#include <algorithm>
+#include <vector>
+
+#include "caffe/layers/lifted_struct_similarity_softmax_layer.hpp"
+#include "caffe/util/math_functions.hpp"
+
+namespace caffe {
+
+template <typename Dtype>
+void LiftedStructSimilaritySoftmaxLossLayer<Dtype>::LayerSetUp(const vector<Blob<Dtype>*>& bottom, const vector<Blob<Dtype>*>& top) {
+  LossLayer<Dtype>::LayerSetUp(bottom, top);
+  CHECK_EQ(bottom[0]->height(), 1);
+  CHECK_EQ(bottom[0]->width(), 1);
+  CHECK_EQ(bottom[1]->channels(), 1);
+  CHECK_EQ(bottom[1]->height(), 1);
+  CHECK_EQ(bottom[1]->width(), 1);
+  // List of member variables defined in /include/caffe/loss_layers.hpp;
+  //   diff_, dist_sq_, summer_vec_, loss_aug_inference_;
+  dist_sq_.Reshape(bottom[0]->num(), 1, 1, 1);
+  dot_.Reshape(bottom[0]->num(), bottom[0]->num(), 1, 1);
+  ones_.Reshape(bottom[0]->num(), 1, 1, 1);  // n by 1 vector of ones.
+  for (int i=0; i < bottom[0]->num(); ++i){
+    ones_.mutable_cpu_data()[i] = Dtype(1);
+  }
+  blob_pos_diff_.Reshape(bottom[0]->channels(), 1, 1, 1);
+  blob_neg_diff_.Reshape(bottom[0]->channels(), 1, 1, 1);
+} 
+
```

Comment:
This layer is not implementing the reshape fiction? I was under the impression this was required for a layer https://github.com/BVLC/caffe/wiki/Development. I guess this will not cause an issue as long as the net is never resized. Context on the addition of the reshape functionality into Caffe https://github.com/BVLC/caffe/pull/594

Reply:
🤷‍♀️ I didn't write this.